### PR TITLE
feat(api): handlers vcs_project 

### PR DESCRIPTION
**Pull request by yesnault**

### CODE REVIEW

File: engine/api/rbac/rule_project.go [hunk 1]
```diff
@@ -2,29 +2,64 @@ package rbac
...
 
 	"github.com/ovh/cds/engine/api/project"
+	"github.com/ovh/cds/engine/cache"
 	"github.com/ovh/cds/sdk"
 )
 
-func ProjectExist(ctx context.Context, db gorp.SqlExecutor, vars map[string]string) error {
+// getProjectIDFromProjectKey returns an id from a project key.
+// this id is stored in cache for 20min
+func getProjectIDFromProjectKey(ctx context.Context, store cache.Store, db gorp.SqlExecutor, projectKey string) (int64, error) {
```
Comment (sguiheux) [inline engine/api/rbac/rule_project.go:17]: should not be in rbac package i think

File: engine/service/http.go [hunk 2]
```diff
@@ -96,6 +103,29 @@ func WriteJSON(w http.ResponseWriter, data interface{}, status int) error {
 	return sdk.WithStack(Write(w, bytes.NewReader(b), status, "application/json"))
 }
 
+// WriteMarshal is a helper function to marshal json, handle errors and set Content-Type for the best
```
Comment (sguiheux) [inline engine/service/http.go:106]:
```suggestion
// WriteMarshal is a helper function to marshal json/yaml, handle errors and set Content-Type for the best
```

File: go.mod [hunk 3]
```diff
@@ -25,6 +24,7 @@ require (
 	github.com/fsamin/go-shredder v0.0.0-20180118184739-b2488aedb5be
 	github.com/fujiwara/shapeio v0.0.0-20170602072123-c073257dd745
 	github.com/gambol99/go-marathon v0.0.0-20170922093320-ec4a50170df7
+	github.com/ghodss/yaml v1.0.0
```
Comment (sguiheux) [inline go.mod:27]: remove yaml tag from sdk new struct

**Review by sguiheux — CHANGES_REQUESTED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#17220.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#17220.1 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

File: engine/api/v2_project_test.go [hunk 4]
```diff
@@ -18,13 +26,125 @@ func Test_addVCSOnProject(t *testing.T) {
...
+	test.NotEmpty(t, uriGet)
+
+	reqGet := assets.NewAuthentifiedRequest(t, u, pass, "GET", uriGet, nil)
+	w2 := httptest.NewRecorder()
+	api.Router.Mux.ServeHTTP(w2, reqGet)
+	require.Equal(t, 200, w2.Code)
 
+	vcsProjects := []sdk.VCSProject{}
+	require.NoError(t, json.Unmarshal(w2.Body.Bytes(), &vcsProjects))
+	require.Len(t, vcsProjects, 1)
```
Comment (fsamin) [inline engine/api/v2_project_test.go:149]: can we add an assertion on the auth field returned ?

File: engine/api/rbac/rule_project.go (hunk 1 above)
Comment (yesnault) [reply engine/api/rbac/rule_project.go:17]: deleted

File: engine/service/http.go (hunk 2 above)
Comment (yesnault) [reply engine/service/http.go:106]: done

File: engine/api/rbac/rule_project.go [hunk 5]
```diff
@@ -2,29 +2,64 @@ package rbac
...
+	var projectID int64
+	var k = cache.Key("projet", "id", "from", "key", projectKey)
```
[the remark this answers is not in the document]
Comment (yesnault) [reply engine/api/rbac/rule_project.go:19]: seen, project_key is now in table rbac_project_keys

File: engine/api/v2_project_test.go (hunk 4 above)
Comment (yesnault) [reply engine/api/v2_project_test.go:149]: done, new field role

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#17228.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✘  
  * Test SDK Rust ✔ 
  * Test UI ✔

File: go.mod (hunk 3 above)
Comment (yesnault) [reply go.mod:27]: done

File: engine/api/rbac/rule_project.go [hunk 6]
```diff
@@ -2,29 +2,64 @@ package rbac
...
 	}
-	if !exist {
-		return sdk.WithStack(sdk.ErrNotFound)
+
+	hasRole, err := HasRoleOnProjectIDAndUserID(ctx, db, role, auth.AuthentifiedUser.ID, projectID)
+	if err != nil {
+		return err
 	}
+
+	log.Debug(ctx, "hasRole:%t projectID:%d role:%s userID:%s", projectID, role, auth.AuthentifiedUser.ID)
```
[the remark this answers is not in the document]
Comment (yesnault) [reply engine/api/rbac/rule_project.go]: field role added

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#17229.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✘  
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#17230.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✘  
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#17233.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#17238.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#17238.1 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#17238.2 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Review by richardlt — APPROVED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#17238.3 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Review by sguiheux — APPROVED** [review]

**Review by fsamin — APPROVED** [review]

**Pull request merged**